src/nfa/operations.cc

```diff
@@ -114,6 +114,364 @@ namespace {
+            if (covered[*i])           // was aready processed
+                continue;
+
+            if (macrostate_vec[*i].IsSubsetOf(check_state)) {
+                covering_set.insert(macrostate_vec[*i]);                // is never covered
+                sub_covering_indexes.push_back(*i);
+            }
+        }
+
+        if (covering_set == check_state) {       // can recurse even without covered :thinking:
+
+            for (unsigned long k = 0; k < sub_covering_indexes.size()-1; k++) {
+                if (macrostate_vec[sub_covering_indexes[k]].size() == 1)            // end on single-sized states
+                    break;
+
+                if (visited[sub_covering_indexes[k]])                               // already processed
+                    continue;
+
+                visited[sub_covering_indexes[k]] = true;
+
+                residual_recurse_coverable(macrostate_vec, sub_covering_indexes, covered, visited, k, subset_map, nfa);
+            }
+
+            covering_set.clear();                 // clear variable to store only needed macrostates
+            for (auto index : sub_covering_indexes) {
+                if (covered[index] == 0) {
+                    auto macrostate_ptr = subset_map->find(macrostate_vec[index]);
+                        if (macrostate_ptr == subset_map->end())        // should never happen
+                             throw std::runtime_error(std::to_string(__func__) + " couldn't find expected element in a map.");
+
+                    covering_set.insert(macrostate_ptr->second);        //todo check for null but should be impossible
+                }
+            }
+
+            remove_covered_state(covering_set, subset_map->find(check_state)->second, nfa);
+            covered[covering_indexes[start_index]] = true;
+        }
+
+
+    }
+
+    Nfa residual_after(const Nfa&  aut) {
+        std::unordered_map<StateSet, State> *subset_map = new std::unordered_map<StateSet,State>();
+        Nfa result;
+        result = determinize(aut, subset_map);
+
+        std::vector <StateSet> macrostate_vec;              // ordered vector of macrostates
+        macrostate_vec.reserve(subset_map->size());
+        for (auto pair: *subset_map) {                   // order by size from largest to smallest
+            macrostate_vec.insert(std::upper_bound(macrostate_vec.begin(), macrostate_vec.end(), pair.first,
+                                [](const StateSet & a, const StateSet & b){ return a.size() > b.size(); }), pair.first);
+        }
+
+        std::vector <bool> covered(subset_map->size(), false);          // flag of covered states, removed from nfa
+        std::vector <bool> visited(subset_map->size(), false);          // flag of processed state
+
+        StateSet covering_set;                // doesn't contain duplicates
+        std::vector<State> covering_indexes;        // indexes of covering states
+        for (unsigned long i = 0; i < macrostate_vec.size()-1; i++) {
```

Comment:
It should be better to use `size_t` type for indices. Then the termination condition variable for `size()` can nicely be extracted, too.